### PR TITLE
linuxPackages.trelay: init at 22.03.5

### PR DESCRIPTION
**Pull request by aprilthepink**
###### Description of changes

Added kernel module trelay.
Trelay is a kernel module that relays ethernet packets between two devices (similar to a bridge), but without any MAC address checks.
This makes it possible to bridge client mode or ad-hoc mode wifi devices to ethernet VLANs, assuming the remote end uses the same source MAC address as the device that packets are supposed to exit from.

Related to https://github.com/NixOS/nixpkgs/issues/241991

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by CutestNekoAqua** [comment]
Continuation of #243569 

@SuperSandro2000 we are adding a own Makefile and not using the openwrt one, cause that assumes that we are building openwrt as a whole and not just the module, and it just does not work in our context. So I opted to instead use a simplistic Makefile, which works and produces a working kernel module. This MAY mean that we have Nix specific bugs (which are reproducible!), but I dont assume it will happen.

File: pkgs/os-specific/linux/trelay/default.nix [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  '';
+
+  makeFlags = [
+    "KERNELRELEASE=${kernel.modDirVersion}"
+    "KERNEL_DIR=${kernel.dev}/lib/modules/${kernel.modDirVersion}/build"
+    "INSTALL_MOD_PATH=$(out)"
+  ];
+
+  meta = with lib; {
+    description = "Trelay relays IP packets between two devices to build a IP bridge between them.";
```
Comment (yu-re-ka) [inline pkgs/os-specific/linux/trelay/default.nix:32]: description is not supposed to start with (or mention) the package name, and should not contain punctuation

File: pkgs/os-specific/linux/trelay/default.nix [hunk 2]
```diff
@@ -0,0 +1,47 @@
+{ stdenv, lib, fetchFromGitHub, kernel, kmod }:
+let
+  version = "22.03.5";
+in
+stdenv.mkDerivation {
+  name = "trelay";
+  version = "${version}-${kernel.version}";
+
+  src = fetchFromGitHub {
```
Comment (yu-re-ka) [inline pkgs/os-specific/linux/trelay/default.nix:9]: The github repo is a mirror. Would prefer the original source.

File: pkgs/os-specific/linux/trelay/default.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
+{ stdenv, lib, fetchFromGitHub, kernel, kmod }:
+let
+  version = "22.03.5";
+in
+stdenv.mkDerivation {
+  name = "trelay";
```
Comment (yu-re-ka) [inline pkgs/os-specific/linux/trelay/default.nix:6]: should be `pname = "trelay";`

File: pkgs/os-specific/linux/trelay/default.nix [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+in
+stdenv.mkDerivation {
+  pname = "trelay";
+  version = "${version}-${kernel.version}";
+
+  src = fetchgit {
+    url = "https://git.openwrt.org/openwrt/openwrt.git";
+    rev = "v${version}";
+    hash = "sha256-zux9BvHt+KXnrZWbpDwWaYg0bxTr49J7teXLDFXbnbU=";
+    sparseCheckout = [ "package/kernel/trelay" ];
```
Comment (yu-re-ka) [inline pkgs/os-specific/linux/trelay/default.nix:13]:
```suggestion
    sparseCheckout = [ "package/kernel/trelay/src" ];
```

**Comment by CutestNekoAqua** [comment]
squashed commits and applied requested changes with Co-author :)

**Comment by yu-re-ka** [comment]
Cross-compilation (`pkgsCross.gnu64.linuxPackages.trelay` on aarch64-linux) fails: https://termbin.com/mvc1
Let me know if you need me to test anything since this requires cross-compiling the entire distro kernel

**Comment by CutestNekoAqua** [comment]
> Cross-compilation (`pkgsCross.gnu64.linuxPackages.trelay` on aarch64-linux) fails: https://termbin.com/mvc1 Let me know if you need me to test anything since this requires cross-compiling the entire distro kernel

Hmm...
> make[1]: *** No rule to make target '/nix/store/s9g19mcasy6y28jayc9mwpwbwd5mkkpd-linux-x86_64-unknown-linux-gnu-6.1.38-dev/lib/modules/6.1.38/source/arch/arm64/Makefile'.  Stop.

I am not that familar with cross compiling, but the logs read like there is a Makefile missing, which seems to not be trelay's, 
If possible can you give me a overview of the directory tree? Something seems wrong here, and that doesnt seem to be a issue with source code

**Comment by CutestNekoAqua** [comment]
Another possible issue: the hash of the git seems to be changed, from when I build it locally to what ofborg seems to get now from openwrt's git? Can someone check if they experience the same issue?

**Comment by yu-re-ka** [comment]
> the hash of the git seems to be changed

This is because we changed the sparseCheckout parameter. You can test this locally with `nix build -f . linuxPackages.trelay.src -vL --rebuild`

**Comment by yu-re-ka** [comment]
> Something seems wrong here, and that doesnt seem to be a issue with source code

It is very likely the Makefile. It tries to search for an `arm64` (which architecture we are building on) directory in the output of the x86 (which we are building for) kernel.

File: pkgs/os-specific/linux/trelay/default.nix [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  sourceRoot = "openwrt/package/kernel/trelay/src";
+  hardeningDisable = [ "pic" "format" ];
+  nativeBuildInputs = [ kmod ] ++ kernel.moduleBuildDependencies;
+
+  postPatch = ''
+    cp '${./Makefile}' Makefile
+  '';
+
+  makeFlags = [
```
Comment (yu-re-ka) [inline pkgs/os-specific/linux/trelay/default.nix:24]:
```suggestion
  makeFlags = kernel.makeFlags ++ [
```

This fixes cross-compilation for me

Comment (CutestNekoAqua) [reply pkgs/os-specific/linux/trelay/default.nix:24]: 🥰

**Comment by CutestNekoAqua** [comment]
I took the bit which was faulty directly out of the wiki. I guess time to improve the docs here!

**Review by yu-re-ka — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Pull request merged**